apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/RetryingReadRowsTest.cs

```diff
@@ -0,0 +1,450 @@
+                        break;
+                }
+            });
+            Assert.Equal(2, rowCount);
+        }
+
+        [Fact]
+        public async Task RetryWithLastScannedKey_Error()
+        {
+            var request = new ReadRowsRequest
+            {
+                Rows = RowSet.FromRowRanges(RowRange.Closed("a", "z"))
+            };
+            var client = Utilities.CreateReadRowsMockClient(
+                request,
+                initialStreamResponse: new[]
+                {
+                    new ReadRowsResponse
+                    {
+                        Chunks =
+                        {
+                            CreateChunk("a", "cf1", "column1", "value1", commitRow: true)
+                        }
+                    },
+                    new ReadRowsResponse
+                    {
+                        LastScannedRowKey = ByteString.CopyFromUtf8("b")
+                    }
+                },
+                responsesForRetryStreams: new[]
+                {
+                    new []
+                    {
+                        new ReadRowsResponse
+                        {
+                            Chunks =
+                            {
+                                CreateChunk("b", "cf1", "column2", "value2", commitRow: true)
+                            }
+                        }
+                    }
+                });
+
+            var stream = client.ReadRows(request);
+
+            // We should be able to read one full row and then there should be an exception in the request validator
+            // of Utilities.CreateReadRowsMockClient because the response form the retry stream with the "b" row key
+            // should be outside the range of (b-z] requested during retry.
+            int rowCount = 0;
+            await Assert.ThrowsAsync<InvalidOperationException>(() => stream.ForEachAsync(row => rowCount++));
+            Assert.Equal(1, rowCount);
+        }
+
+        [Fact]
+        public async Task RetryDuringPartialRow()
+        {
+            var request = new ReadRowsRequest
+            {
+                Rows = RowSet.FromRowRanges(RowRange.Closed("a", "z"))
+            };
+            var client = Utilities.CreateReadRowsMockClient(
+                request,
+                initialStreamResponse: new[]
+                {
+                    new ReadRowsResponse
+                    {
+                        Chunks =
+                        {
+                            CreateChunk("a", "cf1", "column1", "value1")
+                        }
+                    }
+                },
+                responsesForRetryStreams: new[]
+                {
+                    new []
+                    {
+                        new ReadRowsResponse
+                        {
+                            Chunks =
+                            {
+                                CreateChunk("a", "cf1", "column1", "value1"),
```

Comment:
Is this guaranteed not to be (say) "value3"? What would happen in that case?

Reply:
That should be fine. We would reset the state after getting the incomplete row from the first response, so the fact that it was "value1" would be forgotten. This could easily be "value3" and that's what would be seen as the result. I'll add another test for this scenario.

Reply:
Great, thanks - the extra test will definitely help to clarify.